src/plugins/data/common/search/search_source/search_source.ts

```diff
@@ -504,7 +504,17 @@ export class SearchSource {
     const filterSourceFields = (fieldName: string) => {
       return (
         fieldName &&
-        !sourceFiltersValues.some((sourceFilter) => fieldName.match(sourceFilter)) &&
+        !sourceFiltersValues.some((sourceFilter) => {
+          let regexPattern = `^${sourceFilter}`;
+          if (sourceFilter.endsWith('-*')) {
+            regexPattern = `^(${sourceFilter.substring(0, sourceFilter.length - 2)})-*`;
+          } else if (sourceFilter.endsWith('*')) {
+            regexPattern = `^(${sourceFilter.substring(0, sourceFilter.length - 1)})*`;
+          } else if (sourceFilter.startsWith('*') && sourceFilter.length > 1) {
+            regexPattern = `^${sourceFilter.substring(1, sourceFilter.length)}`;
+          }
+          return fieldName.match(regexPattern);
+        }) &&
```

Comment:
This approach has a couple of drawbacks and I think we should exchange the implementation here.

Firstly this allows for regexp injection by the user being able to inject anything into the regexp they want. Given that search source can run on the server, this can lead to a DoS on the Kibana server, by using a ReDoS regexp (see the [ReDoS OWASP](https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS) page).

Secondly the first `if` branch is covered completely by the second one and thus redundant, unless I am missing something.

The Third branch is currently disabling leading wildcards in the field filters. I am not sure if that was on purpose, since  so far we did allow those.

I think we should replace this method with using [`fieldWildcardFilter`](https://github.com/elastic/kibana/blob/master/src/plugins/kibana_utils/common/field_wildcard.ts#L29), which would match the previous logic we had (https://github.com/elastic/kibana/blob/7.11/src/plugins/data/common/search/search_source/search_source.ts#L493-L504) and prevent those issues.

Reply:
As for points 2 and 3:
2. there is a difference between a string ending in `-*` and `*`, they behave differently. so the first if statements covers the first scenario, second if statement convers the second scenario.
3. `^*` is indeed an invalid regex pattern, so I was trying to avoid a scenario in which this occurs

Reply:
The problem is, that currently 1st and 2nd branch are leading to the same. Let's say the filter is `foo-*` it ends in `-*` thus we cut the `-*` away and build then the regex `^foo-*`, by concatenating the ^ + foo + -*. The second branch would detect the same, since `foo-*` ends with `*` and would cut away the last character and add a `*` thus leading to the same `^foo-*`, by concatenating ^ + foo- + * in this case.

The problem with the all of those regexes are, that they are not handling the `*` as a real wildcard, and thus the third one is an invalid regexp. You would want to replace the `*` by a `.*` to match "any character". This logic (in all branches) would still just check that the last character can appear arbitrary often (including none), e.g. a filter of `foo-*` would still match `foobar`, since this matches the regexp `^foo-*` (any word that begins with `foo` and then has an arbitrary amount of hyphens following it (and whatever else behind that)).